torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -4018,6 +4024,69 @@ def rekey_optim_state_dict(
+        """
+        if self.sharding_strategy != ShardingStrategy.NO_SHARD:
+            return None
+        else:
+            return allreduce_hook.allreduce_hook
+
+    def _get_default_comm_hook_state(self) -> Any:
+        r"""
+        Sets a default communication hook state based on a sharding strategy.
+        """
+        if self.sharding_strategy != ShardingStrategy.NO_SHARD:
+            return None
+        else:
+            return allreduce_hook.AllReduceState(process_group=self.process_group)
+
+    def register_comm_hook(self, state: object, hook: callable):
+        """
+        Registers a communication hook which is an enhancement that provides a
+        flexible hook to users where they can specify how FSDP aggregates gradients
+        across multiple workers.
+        This hook can be used to implement several algorithms like GossipGrad
+        and gradient compression which involve different communication strategies for
+        parameter syncs while running Fully Sharded Data Parallel training.
+
+        .. warning::
+            FSDP only support communication hooks for a ``NO_SHARD`` strategy at this time.
+            If other strategies are used, an error will be raised.
+
+        .. warning ::
+            FSDP communication hook should be registered before running an initial forward pass
+            and only once.
+
+        Args:
+            state (object): Passed to the hook to maintain any state information during the training process.
+                            Examples include error feedback in gradient compression,
+                            peers to communicate with next in `GossipGrad <https://arxiv.org/abs/1803.05880>`_, etc.
+                            It is locally stored by each worker
+                            and shared by all the gradient tensors on the worker.
+            hook (callable): Callable with the following signature:
+                            ``hook: Callable[torch.Tensor] -> None``:
+                            This function takes in a python tensor, which represents a gradient
+                            of a loss function with respect to variables of a model
+                            for the local batch that needs to be communicated across ranks.
+                            It then performs all neccessary processing and returns nothing.
+
+        """
+        assert self.check_is_root(), "register_comm_hook can only be called on a root instance."
+        if self.sharding_strategy != ShardingStrategy.NO_SHARD:
+            raise NotImplementedError(
+                "Communication hooks are currently only available for a NO_SHARD strategy."
+            )
+        else:
+            # register same hook for root and all submodules
+            for submodule in self.fsdp_modules(self):
+                assert not submodule._hook_registered, "communication hook can be only registered once"
+                submodule._hook_registered = True
```

Comment:
Curious: is there a way to track this information without an explicit `bool` (i.e. based on the other registered state)?

Reply:
well, as a straightforward suggestion: when a hook is registered, `comm_hook` changes its state from `default` to a `new` one. Technically, we can judge based on current state, i.e. if current is anything, but `default`, we can't register. I do need to think about it more

Reply:
I see. It seems that there is a design choice here that is up to you. Either we can keep the `_hook_registered: bool`, or we can (1) check against default and (2) ensure that `register_comm_hook()` is idempotent for the default hook.

Reply:
Nevermind. I see that @rohan-varma suggested we error if we double-register a hook, even if it is the default one (violating idempotence).